# ATAC-seq pipeline
--------------------

**Preprocessing of the p53 ATAC-seq datasets**
The ATAC-seq data were processed (trimmed, aligned, filtered, and quality controlled) using an in-house the ATAC-seq pipeline*. The model-based analysis of ChIP-seq (MACS2)33 version 2.1.2 was used to identify the peak regions with options -B, -q 0.01â€“nomodel, -f BAM, -g mm. The Irreproducible Discovery Rate (IDR) method34 was used to identify reproducible peaks between two technical replicates . Only peaks reproducible between the two technical replicates were retained for downstream analyses. Peaks for all conditions were then merged together into a standard peak list. The number of raw reads mapped to each standard peak were counted using the intersect function of BedTools35 version 2.26.0. The raw count matrix32 was normalized by Reads Per Million mapped reads (RPM). Pearson correlation coefficients between technical or biological replicates across tissues were calculated based on the Log10 RPM matrix.

*This pipeline is designed for automated end-to-end quality control and processing of ATAC-seq data. The pipeline can be run on compute clusters using parallelized computing with job submission engines as well as on stand alone machines modifying the cluster job part. The pipeline can be run end-to-end, starting from raw FASTQ files all the way to peak calling and signal track generation using a single command [here](MonicaCabreraP/NGS_analysis/ATACseq_analysis). One can also start the pipeline from intermediate stages (for example, using alignment files as input). The pipeline supports paired-end data as well as replicated or non-replicated datasets. Detailed steps can be found [here](MonicaCabreraP/NGS_analysis/ATACseq_analysis)

